stdlib/src/memory/unsafe.mojo

```diff
@@ -1550,6 +1550,188 @@ struct DTypePointer[
+    ](self, offset: SIMD[offset_type, width]) -> SIMD[type, width]:
+        """Gathers a SIMD vector from offsets of the current pointer.
+
+        This method loads from memory addresses calculated by appropriately
+        shifting the current pointer according to the `offset` SIMD vector.
+
+        Constraints:
+            The offset type must be an integral type.
+            The alignment must be 0 or a power of two integer value.
+
+        Parameters:
+            offset_type: The type of the `offset` SIMD vector.
+            width: The SIMD width.
+            alignment: The minimal alignment of the address.
+
+        Args:
+            offset: The SIMD vector of offsets to gather from.
+
+        Returns:
+            The SIMD vector containing the gathered values.
+        """
+        var mask = SIMD[DType.bool, width](True)
+        var default = SIMD[type, width]()
+        return self.gather[width=width, alignment=alignment](
+            offset, mask, default
+        )
+
+    @always_inline("nodebug")
+    fn gather[
+        offset_type: DType,
+        /,
+        *,
+        width: Int = 1,
+        alignment: Int = Self._default_alignment,
+    ](
+        self,
+        offset: SIMD[offset_type, width],
+        mask: SIMD[DType.bool, width],
+        default: SIMD[type, width],
+    ) -> SIMD[type, width]:
+        """Gathers a SIMD vector from offsets of the current pointer.
+
+        This method loads from memory addresses calculated by appropriately
+        shifting the current pointer according to the `offset` SIMD vector,
+        or takes from the `default` SIMD vector, depending on the values of
+        the `mask` SIMD vector.
+
+        If a mask element is `True`, the respective result element is given
+        by the current pointer and the `offset` SIMD vector; otherwise, the
+        result element is taken from the `default` SIMD vector.
+
+        Constraints:
+            The offset type must be an integral type.
+            The alignment must be 0 or a power of two integer value.
+
+        Parameters:
+            offset_type: The type of the `offset` SIMD vector.
+            width: The SIMD width.
+            alignment: The minimal alignment of the address.
+
+        Args:
+            offset: The SIMD vector of offsets to gather from.
+            mask: The SIMD vector of boolean values, indicating for each
+                element whether to load from memory or to take from the
+                `default` SIMD vector.
+            default: The SIMD vector providing default values to be taken
+                where the `mask` SIMD vector is `False`.
+
+        Returns:
+            The SIMD vector containing the gathered values.
+        """
+        constrained[
+            offset_type.is_integral(),
+            "offset type must be an integral type",
+        ]()
+        constrained[
+            (alignment == 0) | _is_power_of_2(alignment),
```

Comment:
why do you allow 0 alignment?

Reply:
Initially, I aligned the implementation of the `gather` and `scatter` methods with the constraints applied to the `alignment` argument in the corresponding LLVM intrinsic functions, which permit a zero value. However, considering the specific context of `DTypePointer` objects, which are always aligned to a non-zero value, I recognize that the flexibility provided by accepting an alignment of 0 may indeed be unnecessary. I have revised the PR accordingly to reflect this more restrictive approach. Do you agree with this change?